src/cls/queue/cls_rgw_queue.cc

```diff
@@ -0,0 +1,664 @@
+            } else if (it->second == info.time) {
+              CLS_LOG(1, "INFO: cls_gc_queue_remove(): erasing tag from urgent data: %s\n", info.tag.c_str());
+              urgent_data.urgent_data_map.erase(info.tag); //erase entry from map, as it will be removed later from queue
+              urgent_data.num_head_urgent_entries -= 1;
+            }
+          }//end-if map end
+        }//end-if urgent data
+        if (! found) {
+          //Search in xattrs
+          bufferlist bl_xattrs;
+          int ret = cls_cxx_getxattr(hctx, "cls_queue_urgent_data", &bl_xattrs);
+          if (ret < 0 && (ret != -ENOENT && ret != -ENODATA)) {
+            CLS_LOG(0, "ERROR: %s(): cls_cxx_getxattrs() returned %d", __func__, ret);
+            return ret;
+          }
+          if (ret != -ENOENT && ret != -ENODATA) {
+            std::unordered_map<string,ceph::real_time> xattr_urgent_data_map;
+            auto iter = bl_xattrs.cbegin();
+            try {
+              decode(xattr_urgent_data_map, iter);
+            } catch (buffer::error& err) {
+              CLS_LOG(1, "ERROR: cls_gc_queue_remove(): failed to decode xattrs urgent data map\n");
+              return -EINVAL;
+            } //end - catch
+            if (xattr_urgent_data_map.size() > 0) {
+              auto found = xattr_urgent_data_map.find(info.tag);
+              if (found != xattr_urgent_data_map.end()) {
+                if (found->second > info.time) {
+                  CLS_LOG(1, "INFO: cls_gc_queue_remove(): tag found in xattrs urgent data map: %s\n", info.tag.c_str());
+                  continue;
+                } else if (found->second == info.time) {
+                  CLS_LOG(1, "INFO: cls_gc_queue_remove(): erasing tag from xattrs urgent data: %s\n", info.tag.c_str());
+                  xattr_urgent_data_map.erase(info.tag); //erase entry from map, as it will be removed later
+                  urgent_data.num_xattr_urgent_entries -= 1;
+                }
+              }
+            } // end - if xattrs size ...
+          } // end - ret != ENOENT && ENODATA
+        }// search in xattrs
+        num_entries++;
+      }//end-for
+      
+      if (num_entries < op.num_entries) {
+        list_op.max = (op.num_entries - num_entries);
+        list_op.start_offset = op_ret.next_offset;
+        out->clear();
+      } else {
+        end_offset = op_ret.offsets[index - 1];
+        CLS_LOG(1, "INFO: cls_gc_queue_remove(): index is %u and end_offset is: %lu\n", index, end_offset);
+        break;
+      }
+    } //end-if
+    else {
+      break;
+    }
+  } while(is_truncated);
+  CLS_LOG(1, "INFO: cls_gc_queue_remove(): Total number of entries to remove: %d\n", total_num_entries);
+
+  if (end_offset != 0) {
+    cls_queue_remove_op rem_op;
+    if (op.marker.empty()) {
+      rem_op.start_offset = 0;
+    } else {
+      rem_op.start_offset = boost::lexical_cast<uint64_t>(op.marker.c_str());
+    }
+
+    rem_op.end_offset = end_offset;
```

Comment:
@cbodley @mattbenjamin : Right now, cls_queue_remove_entries() takes the start offset of first entry and start offset of last entry to be removed. Therefore in that method, the end offset needs to be advanced by the data size by reading the last entry. To avoid reads in cls_queue_remove_entries()for calculating the end offset, can the method just take the start offset of last+1 entry? Is this better or will it confuse the clients?

Reply:
yeah, i think it makes sense for `end_offset` to mean 'one past the end of the range' similar to c++ end() iterators. we just have to make sure that clients can always choose the right end_offset from a listing result - whether that comes from the next entry's offset, or `cls_queue_list_ret::next_offset` if they're removing the last entry in the listing